mcode/rttd-summary-webapp
# Request Configutation File: Save & Upload

Changes include: 

- Load configuration files using the FileReader API 
- Save configuration files using Blob's and a hacky fake-anchor-tag simulated click event 
- Success Alerts that display when the loading or saving of config files is successful 
- Error Alerts that display when loading of config files fails

As you can tell, I went a little overboard with some refactoring along the way. Non-essential changes include: 

- Relocating all settings-panel-specific state and rendering to the `RequestConfigurationPanel.js`
- Renaming the `RequestForm` directory, which used to house all the input components for PatientQuery-related configuration, to `RequestConfiguration`, which now houses all components related to _all_ request configuration.
- Created a `BasicButton.js` file for reusing the same button styling across the new Save and Upload Config buttons. 
- Renames `RequestForm`, which contains all the form data for patient queries, to `PatientQueryForm`. 
- Rename `HeaderForm` to `RequestHeaderForm`, making clear what kind of header we are collecting data for.
- Removed italics from "FHIR Server URL", unifying the label-styling for configuration fields
- Removing the unused FetchDataButton
- Changed the header background color to be zinc-600 instead of slate-600
- Updated the application container to be fluid/full width instead of a max-width container. This helps the application make better usage of limited space on `md-lg` sized screens IMO.
- Changed the single-column breakpoints for the application to be `md:` instead of `sm:`. This helps the application look less smushed on these screens, but definitely encourages us to explore a collapsible menu for the settings panel to improve usability on `sm-md` screens.
- Better unified the icon sizes used in the Request Configuration panel, with all plus buttons going down to `20px`  and all other icons going down to 16px; this is because the plus buttons have an odd amount of padding in the icon art itself. While these sizes are different, it helps better center the icons relative to one another. 


To test:
 - Ensure that previous functionality works as expected (local-host loading, basic RTTD functionality, etc);
 - Ensure that saving a configuration file works;
 - Ensure that loading a configuration file works (try changing fields after saving and ensuring they revert after uploading);
 - Ensure that uploading a garbage file triggers an error alert, but doesn't brick the application (try uploading a FHIR bundlle, for example);
 - Ensure that making requests works with an uploaded configuration file.
